### ISSUE TITLE
upload-artefact@v4 is incompatible with our advertised way of handling multiple python versions

### ISSUE THREAD

**Issue by ewjoachim** [py-cov-action/python-coverage-comment-action#316]
https://github.com/actions/upload-artifact#breaking-changes
> Due to how Artifacts are created in this new version, it is no longer possible to upload to the same named Artifact multiple times. You must either split the uploads into multiple Artifacts with different names, or only upload once. Otherwise you will encounter an error.

We used to upload all coverage files to the same artifact, and they were automatically merged. This is not possible anymore.

**Comment by ewjoachim** [comment]
Well well, seems we're not the only ones with this issue.
https://github.com/actions/upload-artifact/issues/472

Let's wait and see. Meanwhile, if users stumble upon this: we can't really fix it for now, let's wait for GitHub to find a better migration strategy, and meanwhile, please don't upgrade to `@v4`.

**Comment by keunhyung-chung** [comment]
Hello, thanks for developing this awesome action.

I faced the same issue and the new release [`v4.1.0`](https://github.com/actions/download-artifact/releases/tag/v4.1.0) solves this. See https://github.com/actions/upload-artifact/issues/472#issuecomment-1861571655 for more information.

Referring to [this migration guide](https://github.com/actions/download-artifact/blob/v4.1.0/docs/MIGRATION.md), my solution was

```diff
name: CI

on:
  pull_request:
  push:
    branches:
      - "master"
    tags:
      - "*"

jobs:
  build:
    strategy:
      matrix:
        include:
          - python_version: "3.7"
          - python_version: "3.8"
          - python_version: "3.9"
          - python_version: "3.10"

    name: "Python ${{ matrix.python_version }}"
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4

      - name: Set up Python
        id: setup-python
        uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python_version }}

      - name: Install everything, run the tests, produce a .coverage.xxx file
        run: make test # This is the part where you put your own test command
        env:
          COVERAGE_FILE: ".coverage.${{ matrix.python_version }}"
          # Alternatively you can run coverage with the --parallel flag or add
          # `parallel = True` in the coverage config file.
          # If using pytest-cov, you can also add the `--cov-append` flag
          # directly or through PYTEST_ADD_OPTS.

      - name: Store coverage file
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
        with:
-          name: coverage
+          name: coverage-${{ matrix.python_version }}
           path: .coverage.${{ matrix.python_version }}

  coverage:
    name: Coverage
    runs-on: ubuntu-latest
    needs: build
    permissions:
      pull-requests: write
      contents: write
    steps:
      - uses: actions/checkout@v4

-      - uses: actions/download-artifact@v3
+      - uses: actions/download-artifact@v4
        id: download
        with:
-          name: "coverage"
+          pattern: coverage-*
+          merge-multiple: true

      - name: Coverage comment
        id: coverage_comment
        uses: py-cov-action/python-coverage-comment-action@v3
        with:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          MERGE_COVERAGE_FILES: true

      - name: Store Pull Request comment to be posted
        uses: actions/upload-artifact@v3
        if: steps.coverage_comment.outputs.COMMENT_FILE_WRITTEN == 'true'
        with:
          name: python-coverage-comment-action
          path: python-coverage-comment-action.txt
```

**Comment by ewjoachim** [comment]
Yep, that's the issue I linked, and you're right: there now is a solution.

I like that they spend the time checking that this fix is also re-run compatible.

That said: artifacts are used in 2 places in the action: for the part you fixed, and also for saving the comment to be published in case the PR is from a fork (so the PR workflow doesn't have enough permissions to post a comment to the PR and we need to split that to a different workflow).

We need to check if that second step can be switched to v4 as-is . All in all, it's always good to test for real, but our automated tests should be enough to figure that out, so as long as we also update [this line](https://github.com/py-cov-action/python-coverage-comment-action/blob/main/tests/end_to_end/repo/.github/workflows/ci.yml#L47), I'm good.

Is it something you'd like to do ? If so, feel free to submit a PR :)
- Update the README with the instructions
- Update the end to end test (linked above)
- Update the versions used by this repo itself, though it's independant
- Report and update any other uses of `upload-artifact` or `download-artifact`

> Hello, thanks for developing this awesome action.

Thanks to you for using it :heart:, and happy to help, sharing the kudos with kieferro :tada:

**Comment by keunhyung-chung** [comment]
> Yep, that's the issue I linked, and you're right: there now is a solution.
> 
> I like that they spend the time checking that this fix is also re-run compatible.
> 
> That said: artifacts are used in 2 places in the action: for the part you fixed, and also for saving the comment to be published in case the PR is from a fork (so the PR workflow doesn't have enough permissions to post a comment to the PR and we need to split that to a different workflow).
> 
> We need to check if that second step can be switched to v4 as-is . All in all, it's always good to test for real, but our automated tests should be enough to figure that out, so as long as we also update [this line](https://github.com/py-cov-action/python-coverage-comment-action/blob/main/tests/end_to_end/repo/.github/workflows/ci.yml#L47), I'm good.
> 
> Is it something you'd like to do ? If so, feel free to submit a PR :)
> 
> * Update the README with the instructions
> * Update the end to end test (linked above)
> * Update the versions used by this repo itself, though it's independant
> * Report and update any other uses of `upload-artifact` or `download-artifact`
> 
> > Hello, thanks for developing this awesome action.
> 
> Thanks to you for using it ❤️, and happy to help, sharing the kudos with kieferro 🎉

@ewjoachim Thanks, I have created PR #324. I will open this after it is ready.

**Comment by keunhyung-chung** [comment]
@ewjoachim 
#324 is ready to review.

**Comment by ewjoachim** [comment]
Fixed by #324 / #326
